### PR TITLE
New global search

### LINKED ISSUE #4096: Global search not available
Currently that the content of the search field is bound to a tab. In the old versions, it was globally stored. So I could click through the tabs and see if an entry is available somewhere.

search

Refs https://github.com/JabRef/jabref/pull/1549, https://github.com/JabRef/jabref/issues/104#issuecomment-157304997, https://github.com/JabRef/jabref/pull/332, https://github.com/JabRef/jabref/pull/1981

This is the thing, I miss most. I think, we described all our search features at https://help.jabref.org/en/Search for a description of our search capabilities. I can agree to go through the documentation and redesign it.

### PR DESCRIPTION
**Pull request by Siedlerchr**
Fixes #4096 

Experimental implementation of the global search.
- Planned followups should include some sort of jump-to-entry-on-click
- [x] Keep search string when selecting different tabs
- [x] Entry preview

grafik

__

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by calixtus** [comment]
@ytzemih Great news: @Siedlerchr was able to bring back the GlobalSearchDialog.

**Comment by ytzemih** [comment]
@calixtus @Siedlerchr Awesome news. Thanks a lot for reintroducing that feature!

File: src/main/java/org/jabref/gui/StateManager.java [hunk 1]
```diff
@@ -128,6 +131,18 @@ public void setSearchQuery(SearchQuery searchQuery) {
         activeSearchQuery.setValue(Optional.of(searchQuery));
     }
 
+    public void setGlobalSearchActive(boolean active) {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/StateManager.java:134]: Do you really need to have this as a state variable? It seems like the only place where you need to know if the search bar is open already open is if the user starts a new search, so this seems to be a very localized thing.

File: src/main/java/org/jabref/gui/maintable/MainTableDataModel.java [hunk 2]
```diff
@@ -51,10 +51,10 @@ public MainTableDataModel(BibDatabaseContext context, PreferencesService prefere
         IntegerProperty resultSize = new SimpleIntegerProperty();
         resultSize.bind(Bindings.size(entriesFiltered));
         stateManager.setActiveSearchResultSize(context, resultSize);
+        stateManager.setActiveSearchResultSize(context, resultSize);
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/maintable/MainTableDataModel.java:54]: duplicate?

File: src/main/java/org/jabref/gui/DialogService.java [hunk 3]
```diff
@@ -168,6 +168,14 @@ boolean showConfirmationDialogWithOptOutAndWait(String title, String content,
      */
     void showCustomDialog(BaseDialog<?> dialog);
 
+    /**
+     * This will create and display a new dialog showing a custom {@link DialogPane}
+     * and using custom {@link ButtonType}s.
+     *
+     *
+     */
+    Optional<ButtonType> showNonModalCustomDialogAndWait(String title, DialogPane contentPane, ButtonType... buttonTypes);
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/DialogService.java:177]: Maybe add the modality as a new optional enum to the already existing showCustomDialogAndWait method?

File: src/main/java/org/jabref/gui/search/GlobalSearchBar.java [hunk 4]
```diff
@@ -296,15 +312,11 @@ private void informUserAboutInvalidSearchQuery() {
 
     public void setAutoCompleter(SuggestionProvider<Author> searchCompleter) {
         if (preferencesService.getAutoCompletePreferences().shouldAutoComplete()) {
-            AutoCompletionTextInputBinding<Author> autoComplete = AutoCompletionTextInputBinding.autoComplete(searchField,
-                    searchCompleter::provideSuggestions,
-                    new PersonNameStringConverter(false, false, AutoCompleteFirstNameMode.BOTH),
-                    new AppendPersonNamesStrategy());
+            AutoCompletionTextInputBinding<Author> autoComplete = AutoCompletionTextInputBinding.autoComplete(searchField,searchCompleter::provideSuggestions,                                                                                       new AppendPersonNamesStrategy());
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/search/GlobalSearchBar.java:315]: Why did you remove the  PersonNameStringConverter here?

File: src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    public void doGlobalSearch() {
+        if (stateManager.isGlobalSearchActive()) {
+
+            BibDatabaseContext resultDbContext = new BibDatabaseContext();
+
+            for (BibDatabaseContext dbContext : this.stateManager.getOpenDatabases()) {
+
+                List<BibEntry> result = dbContext.getEntries().stream().filter(entry -> isMatched(stateManager.activeGroupProperty(), stateManager.activeSearchQueryProperty().get(), entry))
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java:65]: better pass the search query as a parameter? Feels cleaner in my opinion, and you don't rely on the fact that doGlobalSearch needs to be called after the state property is updated.

Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java:65]: Does it really make sense to filter the entries by the active group for the global search? I would expect that this would result in a very counter-intuitive behavior.

File: src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+                                                 .map(currEntry -> {
+                                                     BibEntry newEntry = currEntry.withField(new UnknownField(GlobalSearchResultDialog.LIBRARY_NAME_FIELD), FileUtil.getBaseName(dbContext.getDatabasePath().orElse(null)));
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java:67]: This is a very ugly hack, and if I'm not mistaken it even modifies the existing entry. Please try to find a different solution, like using a List of pairs of entry + dbcontext.

File: src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+import org.jabref.logic.util.io.FileUtil;
+import org.jabref.model.database.BibDatabaseContext;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.field.UnknownField;
+import org.jabref.model.groups.GroupTreeNode;
+import org.jabref.model.search.matchers.MatcherSet;
+import org.jabref.model.search.matchers.MatcherSets;
+import org.jabref.preferences.PreferencesService;
+
+public class GlobalSearchResultDialog {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java:33]: Why not make this a real dialog like all the other dialogs we have? That is, derive from Dialog, have fxml etc

File: src/main/java/org/jabref/gui/search/SearchResultsTable.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+import org.jabref.gui.maintable.MainTable;
+import org.jabref.gui.maintable.MainTableColumnFactory;
+import org.jabref.gui.maintable.MainTableDataModel;
+import org.jabref.gui.maintable.MainTablePreferences;
+import org.jabref.gui.maintable.SmartConstrainedResizePolicy;
+import org.jabref.gui.maintable.columns.MainTableColumn;
+import org.jabref.model.database.BibDatabaseContext;
+import org.jabref.preferences.PreferencesService;
+
+public class SearchResultsTable extends TableView<BibEntryTableViewModel> {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/search/SearchResultsTable.java:21]: As this class has quite a bit of code duplication with the main table, can we have one class that handles the display of entries (in a table). Would make it easier to reuse it.

File: src/main/java/org/jabref/gui/search/SearchResultsTable.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+                    .map(column -> (MainTableColumn<?>) column)
+                    .filter(column -> column.getModel().equals(columnModel))
+                    .findFirst()
+                    .ifPresent(column -> this.getSortOrder().add(column)));
+
+        if (mainTablePreferences.getResizeColumnsToFit()) {
+            this.setColumnResizePolicy(new SmartConstrainedResizePolicy());
+        }
+
+        this.getSelectionModel().setSelectionMode(SelectionMode.MULTIPLE);
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/search/SearchResultsTable.java:55]: What happens if an entry is selected?

**Review by tobiasdiez — CHANGES_REQUESTED** [review]

File: src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java (hunk 5 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java:73]: Good point

File: src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java (hunk 7 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java:36]: In the making.

Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java]: in progress

File: src/main/java/org/jabref/gui/search/SearchResultsTable.java (hunk 9 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/search/SearchResultsTable.java:55]: preview will be shown, will change to single

Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/search/SearchResultsTable.java]: Wouldn't it be more intuitive if the entry is selected in the main table?

File: src/main/java/org/jabref/gui/search/SearchResultsTable.java (hunk 8 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/search/SearchResultsTable.java:21]: I know, but it's just part of the constructor and this is exactly what the want. Show the user a table he's familiar with

Comment (calixtus) [reply src/main/java/org/jabref/gui/search/SearchResultsTable.java:21]: I think this would escalate this PR way too much. Probably something for a later code quality PR.

File: src/main/java/org/jabref/gui/DialogService.java (hunk 3 above)
Comment (calixtus) [reply src/main/java/org/jabref/gui/DialogService.java]: As this addition is doubted, in the current state of implementation of this feature this is not used anymore and we want to finish this PR I removed it. Maybe we can add a modal enum in a later PR when we need it again.

**Review by calixtus — APPROVED** [review]

**Review by koppor — APPROVED** [review]
